packages/astro/src/core/module-loader/vite.ts

```diff
@@ -102,11 +105,18 @@ export function createViteLoader(viteServer: vite.ViteDevServer): ModuleLoader {
 			});
 		},
 		webSocketSend(msg) {
-			return viteServer.hot.send(msg);
+			return ssrEnvironment.hot.send(msg);
 		},
 		isHttps() {
-			return !!viteServer.config.server.https;
+			return !!ssrEnvironment.config.server.https;
 		},
 		events,
 	};
 }
+
+export function getRunnableEnvironment(viteDevServer: ViteDevServer): RunnableDevEnvironment {
+	if (isRunnableDevEnvironment(viteDevServer.environments.ssr)) {
+		return viteDevServer.environments.ssr;
+	}
+	throw new Error("The environment isn't a runnable dev environment.");
```

Comment:
Ah we missed each other. Same question: what happens in a non-runnable dev enviornment?

Reply:
If I interpreted the docs correctly, non-runnable dev environments are:
- client
- environments different from the vite ones

This guard should prevent passing environments that cannot run modules. I added a simple runtime error to catch the misuse early during development.